scala/ScalaWorker.scala

```diff
@@ -0,0 +1,220 @@
+    }
+
+    // Options copied from Nailgun.scala in Zinc
+    val options = List("-cp", classpath, "-server", "-Xms1024m", "-Xmx3g", "-XX:MaxPermSize=384m",
+      "-XX:ReservedCodeCacheSize=192m")
+    val cmd = "java" :: options ++ Seq(classOf[Nailgun].getName, s"$zincPort")
+    val builder = Process(cmd)
+    this.nailgunProcess = builder.run(logger)
+
+    serverArgs = cmd.mkString(" ")
+    zincClient = new ZincClient(port = zincPort)
+  }
+
+  private def awaitServer() {
+    var count = 0
+    while (!zincClient.serverAvailable && (count < 50)) {
+      try { Thread.sleep(100) } catch { case _: InterruptedException => }
+      count += 1
+    }
+  }
+
+  def main(args: Array[String]): Unit = {
+    if (args.contains("--persistent_worker")) {
+      startServer(args(0))
+      runPersistentWorker(args)
+    } else {
+      // This is a single invocation of the example that exits after it processed the request.
+      ZincMain.run(args, cwd = None)
+    }
+  }
+
+  private def listFiles(f: File): Seq[String] = {
+    val current = f.listFiles
+    val files = current.filter(_.isFile).map(_.getAbsolutePath)
+    val directories = current.filter(_.isDirectory)
+    files ++ directories.flatMap(listFiles)
+  }
+
+  // Extract a src jar to a temporary directory and return the list of extracted files
+  private def expandSrcJar(path: String): Seq[String] = {
+    val tempDir = Files.createTempDirectory(null).toFile
+    Seq("unzip", "-q", path, "-d", tempDir.getAbsolutePath).!!
+    listFiles(tempDir)
+  }
+
+  @throws[IOException]
+  private def runPersistentWorker(args: Array[String]) {
+    val originalStdOut = System.out
+    val originalStdErr = System.err
+
+    while (true) {
+      try {
+        val request = WorkRequest.parseDelimitedFrom(System.in)
+        if (request == null) {
+          return
+        }
+
+        inputs.clear()
+
+        for (input <- request.getInputsList().asScala) {
+          inputs.put(input.getPath(), input.getDigest().toStringUtf8())
+        }
+
+        val baos = new ByteArrayOutputStream()
+        var exitCode = 0
+
+        val ps = new PrintStream(baos)
+        try {
+          System.setOut(ps)
+          System.setErr(ps)
+
+          var clientArgs: Seq[String] = null
+
+          try {
+            clientArgs = request.getArgumentsList.asScala.flatMap { arg =>
+              // srcjars must be extracted before we can pass them to zinc
+              if (arg.endsWith(".srcjar")) {
+                expandSrcJar(arg)
+              } else {
+                Seq(arg)
+              }
+            }
+            awaitServer()
+            exitCode = zincClient.run(
+              args = clientArgs,
+              cwd = new File(System.getProperty("user.dir")),
+              out = ps,
+              err = ps
+            )
+          } catch {
+            case e: Exception =>
+              // We use System.out.println as not to accidentally write to real stdout
```

Comment:
I don't follow this comment. How is it not real stdout?

Can we not use stderr here?